design/1495-centralized-venvs.md

```diff
@@ -0,0 +1,76 @@
+## Scope
+
+This proposal contemplates a truly minimal MVP with the intention of getting the core feature
+shipped soon so any additional features/discussions can be based on actual usage patterns/needs.
+
+
+## Definitions
+
+* `$cache_dir` represents the directory where uv stores it's cache.  Any mechanisms currently in
+   place for setting the directory, e.g. `UV_CACHE_DIR` and `XDG_CACHE_DIR`, remain unchanged.
+* `$venv_cache_dir`: `$cache_dir/venvs-v0`
+* `$uv_venv_dir`: the file system path to UV's venv location, regardless of how it is determined
+
+## Proposal
+
+1. Document this feature as a "preview" for ~six months to make it easier to facilitate backwards
+  incompatible changes should real-world usage patterns indicate they are needed.
+   - `--preview-features venvs-in-cache` to silence warning
+1. Application
+   - Only applies to project based venvs
+   - Does NOT apply to `uv venv` invocations where a PATH is provided
+1. venv storage location:
+   - a cache bucket dedicated to venvs: `$venv_cache_dir`
+   - customizable: no
+1. Trigger centralized venv storage when:
+   - `UV_VENVS_IN_CACHE`: is set to any value;
+   - OR `venvs-in-cache` setting is `true`
+1. Venv dir name calculation
+   - [ChatGTP input on this
+     question](https://chatgpt.com/share/68766176-ca84-8010-84a3-b8a3d21f4b32), I'm following it's
+     recommendation
+   - Human‑readable slug + short hash (pseudocode below)
+     ```
+      slug = slugify(basename(project_root))
+      hash = sha256(project_root)[0:8]
+      venv_dir = $venv_cache_dir/{slug}-{hash}
+     ```
+1. Place text file in the venv which contains the path to the uv project the venv was created for:
+   - Path: `$venv_dir/uv-project-path.txt`
+   - Contents example: `/home/joe/projects/foo`
+   - Note: the directory is the project directory, not the path to the `pyproject.toml` or `uv.toml`
+1. Add `uv venv --dir` to output the absolute normlized path to the current venv location
+   - This is not limited to the venv-in-cache feature
+   - E.g.: `print($uv_venv_dir)`
+1. Cache cleanup
+   - `uv cache clean` removes all venvs
+
+
+## Out of Scope / Rejections
+
+1. We will not support full-paths inside the `$cache_dir/venvs-v0`
+   - Avoid possibility of hitting path length limits
+   - Doesn't match what uv currently does
+   - "Flat is better than nested"
+2. Using a symlink in the venv to point back to the project directory
+   - Windows doesn't have a good symlink solution
```

Comment:
I think you'll want to look at how pnpm approaches this (maybe not for this proposal, but in general)

They symlink the whole `node_modules` directory - and this also works on windows. 
No clue HOW they do it exactly - but i think it'd allow for a global cache of installed modules - greatly reducing duplicate files on the system.

Reply:
It would appear uv already does this:

- https://github.com/astral-sh/uv/issues/11504
- https://docs.astral.sh/uv/reference/settings/#link-mode

Reply:
@eabase please understand, those are rejected ideas in that section.

> How, then, would you use multiple project directories sharing a venv?

We wouldn't.  This proposal explicitly identifies sharing venvs as out of scope.  I understand you have a use case for this, and care a lot about it, we don't consider that use case valid for this proposal.

Reply:
I've not looked into how uv handles the shared cache in Windows.  If the symlink equivalent works there, then having a symlink back to the project directory is feasible.  

I'd still argue a text file with the path as it's content is simpler and has less edge cases and therefore is the more appropriate solution.

Reply:
I'm not even arguing the decision to not symlink back to the project. This may facilitate _shared_ environments (maybe in the future).

I just think it should be argued better - not with "because windows" - where the problem has actually been solved _within the same project_.

Reply:
Fair enough.  If we get some indication from the project maintainers that they are likely to move on this proposal, I'll consider updating the doc at that point.